File: 0303-range-sum-query-immutable/0303-range-sum-query-immutable.js
/**
 * @param {number[]} nums
 */
var NumArray = function(nums) {
    this.prefixSums = PrefixSum(nums);
};
var PrefixSum = function(nums) {
    let res = []
    return nums.reduce((acc, curr, index) => {
          if (acc.length === 0) {
             acc.push(0)
          }
          
          acc.push(acc[index] + curr)
         return acc
    },[])

}

/** 
 * @param {number} left 
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function(left, right) {
    // prefix[right + 1] - prefix[left];
    return this.prefixSums[right + 1] - this.prefixSums[left]
};

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */

File: 0303-range-sum-query-immutable/0303-range-sum-query-immutable.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.nums = nums  

    def sumRange(self, left: int, right: int) -> int:
        prefix = [0]
        for item in self.nums:
            prefix.append(prefix[len(prefix) -1] + item)
        return prefix[right+1] - prefix[left]

# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)